### PR TITLE
[RELAND] Add __torch_function__ benchmarks

### PR DESCRIPTION
**Pull request by hameerabbasi**
Re-land of #35530 and #34645

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
waiting on CI. For reference, can you cross reference the previous time we landed this PR?

**Comment by hameerabbasi** [comment]
Out of the three failures:

* Bazel is broken in upstream, and there's a merge conflict if I do not merge that commit.
* `pytorch_linux_xenial_py3_clang5_asan_test`: Introduces UB, but I'm not touching any C/C++ code.
* The `pr/caffe2-py3.6-devtoolset7-rocmrpm-centos7-test` failure is a timeout:

```
14:13:52 [ RUN      ] UtilityOpGPUTest.testReshapeWithScalar
17:12:51 Build timed out (after 180 minutes). Marking the build as failed.
```

**Comment by ezyang** [comment]
The ASAN failure is very clearly related to the code you added though, because it's reporting UB when running bench.py. You're probably dividing by zero from userland and that is triggering UBSAN...

File: benchmarks/overrides_benchmark/bench.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+def bench(t1, t2):
+    bench_times = []
+    for _ in range(NUM_REPEAT_OF_REPEATS):
+        time_start = time.time()
+        for _ in range(NUM_REPEATS):
+            torch.add(t1, t2)
+        bench_times.append(time.time() - time_start)
+
+    bench_time = float(torch.min(torch.Tensor(bench_times))) / 1000
+    bench_std = float(torch.std(torch.Tensor(bench_times))) / 1000
```
Comment (hameerabbasi) [inline benchmarks/overrides_benchmark/bench.py:20]: @ezyang Only division happens here, and it's by 1000, not 0. Elsewhere I'm only doing `torch.add`.

Comment (hameerabbasi) [reply benchmarks/overrides_benchmark/bench.py:20]: Got it, `torch.std` has UB with a length-1 Tensor.

**Comment by ezyang** [comment]
If you think std shouldn't UB when passed a one length tensor (a view I'm sympathetic with), file a bug report!

**Pull request closed without merging**